enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,821 @@
+`AAAA` record exists for the DNS name.
+
+### User Stories
+
+* As an OpenShift cluster administrator, I want to add regular DNS names to EgressFirewall rules, so that I can allow
+access to them even if the IP addresses associated with the corresponding DNS records change. 
+* As an OpenShift cluster administrator, I want to add wildcard DNS names to EgressFirewall rules, so that I can
+allow access to all the subdomains belonging to the wildcard DNS names.
+
+### Goals
+
+* Support allowing access to DNS names even if the IP addresses associated with them changes
+* Support usage of wildcard DNS names in EgressFirewall rules.
+
+### Non-Goals
+
+* Support additional DNS resolution functionality in the new CoreDNS plugin. The new plugin will only inspect the
+response of DNS resolution by the other existing plugins.
+* Support denying of DNS lookups in the new CoreDNS plugin. The new plugin will not stop the DNS lookup itself and will not
+respond with a `REFUSED`/`NXDOMAIN` response code if a EgressFirewall rule denies access to the specific DNS name being queried for.
+
+## Proposal
+
+This enhancement proposes to introduce a new CoreDNS [external plugin](https://coredns.io/explugins/) (`ocp_dnsnameresolver`) and a new Custom Resource
+(`DNSNameResolver`) to improve the integration of CoreDNS with EgressFirewall. This proposal takes the OVN Interconnect (OVN-IC)
+architecture into consideration where there are (possibly) multiple OVN-K masters and a centralized OVN-K cluster manager. The OVN-K cluster manager will
+create a `DNSNameResolver` CR for each unique DNS name (both regular and wildcard DNS names) used in
+the EgressFirewall rules. The same DNS name may appear in multiple EgressFirewall rules, but only a single `DNSNameResolver` CR object will be used
+for a unique DNS name. This CR will be used to store the DNS name along with the current IP addresses, the corresponding
+TTL, and the last lookup time. The `DNSNameResolver` CR is meant for communication between
+CoreDNS and OVN-K master(s).
+
+The new plugin will inspect each DNS lookup and the corresponding response for the DNS lookup from other
+plugins. If the DNS name in the query matches any `DNSNameResolver` CR(s) (regular or wildcard or both), then the
+plugin will update the `.status` of the matching `DNSNameResolver` CR(s) with the DNS name along with the IP addresses,
+the corresponding TTL, and the last lookup time. The OVN-K master(s) will watch the `DNSNameResolver`
+CRs. Whenever the IP addresses are updated for a `DNSNameResolver` CR, the OVN-K master(s) will update the underlying `AddressSet`
+referenced by the ACL rule(s) for the corresponding EgressFirewall rule(s).
+
+A new controller (`DNSNameResolver` controller) will keep track of the next lookup time (TTL + last lookup time)
+for each regular DNS name and send a DNS lookup
+query to CoreDNS when the minimum TTL expires. However, for a wildcard DNS name a DNS lookup cannot be only performed
```

Comment:
What is the expected minimum TTL?  If it is the same as (or longer than) the rate at which EgressFirewall formerly updated the records (30m?) , what improvement is gained?  If it's more frequent, are we going to be peppering CoreDNS with lookup requests that might not need to be updated?  What if someone mistakenly set the TTL to 1s?

Reply:
> What is the expected minimum TTL?

The minimum TTL will vary for each DNS names. For a DNS name, the associated IP addresses may have different TTLs. The minimum TTL among these will be used to set the next lookup time of the DNS name. 

> If it is the same as (or longer than) the rate at which EgressFirewall formerly updated the records (30m?) , what improvement is gained? If it's more frequent, are we going to be peppering CoreDNS with lookup requests that might not need to be updated?

The 30m interval is used as a default TTL. If the minimum TTL of a DNS name is less than 30m, then that value will be used for the next lookup.

There is almost no change in behavior regarding the frequency of lookups based on the TTL as compared how it is currently done by EgressFirewall. The only changes are that the default TTL of 30m is not used when the minimum TTL value of a DNS name is higher than 30m, and when the minimum TTL of a DNS name is zero, instead of using the default TTL of 30m, a TTL of 5s is used.

> What if someone mistakenly set the TTL to 1s?

Currently, it will be looked up at a frequency of 1s (both in the existing implementation and the proposed design). We can probably use 5s as the minimum TTL for any TTL value less than 5s.